File: Programming/02. CSharp Part 2/08.StringsTextProcessing/03.CheckTheBrackets/CheckTheBrackets.cs
//Write a program to check if in a given expression the brackets are put correctly.
//Example of correct expression: ((a+b)/5-d).
//Example of incorrect expression: )(a+b)).

using System;
using System.Collections.Generic;
class CheckTheBrackets
{
    static void Main()
    {
        Console.Write("Enter the math expression: ");
        string expression = Console.ReadLine();
        Stack<char> bracketsStack = new Stack<char>();

        // loop trough the elements of the string
        for (int charIndex = 0; charIndex < expression.Length; charIndex++)
        {
            // if a openning bracke is found
            if (expression[charIndex] == '(')
            {
                // push that bracket to the stack
                bracketsStack.Push(expression[charIndex]);
            }
                // if closing bracket is found
            else if (expression[charIndex] == ')')
            {
                // if the stack is empty then the expression is wrong;show the msg and break the loop
                if (bracketsStack.Count == 0)
                {
                    // adding the last char so that the stack wont be empty
                    bracketsStack.Push(expression[charIndex]);
                    // break the loop with not empty stack -> will counse the wrong brackets message 
                    break;
                }
                // take out a ) bracket
                bracketsStack.Pop();
            }
        }

        // if the stack is empty then the expression is correct
        if (bracketsStack.Count == 0)
        {
            Console.WriteLine("The expression has well placed brackets!");
        }
            // else the expression is wrong or the brackets are missplaces
        else
        {
            Console.WriteLine("Wrong brackets position!");
        }
    }
}

File: HTML and CSS/JS/01.JS Fundamentals/11.Strings/js/checkTheBrackets.js
//Write a JavaScript function to check if in a given expression the brackets are put correctly.
//Example of correct expression: ((a+b)/5-d).
//    Example of incorrect expression: )(a+b)).


function onBracketsCheck() {
    printInElement('', true);
    var expression = "((a+b)/5-d))";

    printInElement("The expression '"+ expression + "'");
    if (isCorrectExpression(expression)) {
        printInElement(" is correct!");
    } else {
        printInElement(" is NOT correct!");
    }
}

function isCorrectExpression(expression) {
    var correct = 0;
    var stringExpression = new String(expression);
    for (var i = 0; i < stringExpression.length; i++) {
        if (stringExpression[i] === ')') {
            correct--;
        } else if (stringExpression[i] == '(') {
            correct++;
        }
        if (correct < 0) {
            return false;
        }
    }

    if (correct === 0) {
        return true;
    } else {
        return false;
    }
}